File: apps/android/app/src/main/java/com/exponential/app/data/api/TeamsApi.kt
package com.exponential.app.data.api

import com.exponential.app.data.db.TeamEntity
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.Serializable

@Serializable
data class EnsureDefaultResult(val team: TeamEntity)

@Serializable
data class DeleteTeamInput(val teamId: String)

@Serializable
data class DeleteBoardInput(val boardId: String)

@Serializable
private object EmptyInput

@Singleton
class TeamsApi @Inject constructor(private val trpc: TrpcClient) {

    suspend fun ensureDefault(accountId: String): TeamEntity =
        trpc.mutation(
            accountId,
            path = "teams.ensureDefault",
            input = EmptyInput,
            inputSerializer = EmptyInput.serializer(),
            outputSerializer = EnsureDefaultResult.serializer(),
        ).team

    suspend fun delete(accountId: String, teamId: String) {
        trpc.mutationUnit(
            accountId,
            path = "teams.delete",
            input = DeleteTeamInput(teamId),
            inputSerializer = DeleteTeamInput.serializer(),
        )
    }

    suspend fun deleteBoard(accountId: String, boardId: String) {
        trpc.mutationUnit(
            accountId,
            path = "boards.delete",
            input = DeleteBoardInput(boardId),
            inputSerializer = DeleteBoardInput.serializer(),
        )
    }
}

File: apps/ios/ExpCore/Sources/API/TeamsApi.swift
import Foundation

public struct GetDefaultResult: Decodable, Sendable {
    // Nullable: teams.getDefault never creates (EXP-188) — a fresh signup
    // (or an owner who deleted their last team) resolves to nil.
    public let team: TeamResult?

    public init(team: TeamResult?) {
        self.team = team
    }
}

public struct CreateTeamResult: Decodable, Sendable {
    public let team: TeamResult

    public init(team: TeamResult) {
        self.team = team
    }
}

public struct TeamResult: Decodable, Sendable {
    public let id: String
    public let name: String
    public let slug: String

    public init(id: String, name: String, slug: String) {
        self.id = id
        self.name = name
        self.slug = slug
    }
}

public struct CreateTeamInput: Encodable, Sendable {
    public let name: String
    public var iconUrl: String?

    public init(name: String, iconUrl: String? = nil) {
        self.name = name
        self.iconUrl = iconUrl
    }
}

public struct DeleteTeamInput: Encodable, Sendable {
    public let teamId: String

    public init(teamId: String) {
        self.teamId = teamId
    }
}

public struct DeleteBoardInput: Encodable, Sendable {
    public let boardId: String

    public init(boardId: String) {
        self.boardId = boardId
    }
}

public final class TeamsApi: Sendable {
    private let trpc: TrpcClient

    public init(trpc: TrpcClient) {
        self.trpc = trpc
    }

    /// The NON-CREATING default-team resolver (EXP-188, replaced
    /// `teams.ensureDefault`): the user's oldest non-feedback membership team,
    /// or nil when they have none. An input-less tRPC query, so it goes over
    /// GET like the other reads.
    public func getDefault(accountId: String) async throws -> TeamResult? {
        let result: GetDefaultResult = try await trpc.query(accountId: accountId, path: "teams.getDefault")
        return result.team
    }

    public func create(accountId: String, name: String, iconUrl: String? = nil) async throws -> TeamResult {
        let result: CreateTeamResult = try await trpc.mutation(accountId: accountId, path: "teams.create", input: CreateTeamInput(name: name, iconUrl: iconUrl))
        return result.team
    }

    public func delete(accountId: String, teamId: String) async throws {
        try await trpc.mutationVoid(accountId: accountId, path: "teams.delete", input: DeleteTeamInput(teamId: teamId))
    }

    public func deleteBoard(accountId: String, boardId: String) async throws {
        try await trpc.mutationVoid(accountId: accountId, path: "boards.delete", input: DeleteBoardInput(boardId: boardId))
    }
}
